hazelcast/src/main/java/com/hazelcast/util/ExceptionUtil.java

```diff
@@ -48,23 +48,39 @@ public static String toString(Throwable cause) {
         return sw.toString();
     }
 
+    public static RuntimeException peel(final Throwable t) {
+        return (RuntimeException) peel(t, null);
+    }
+
+    public static <T extends Throwable> Throwable peel(final Throwable t, Class<T> allowedType) {
+        if (t instanceof RuntimeException) {
+            return t;
+        }
+
+        if (t instanceof ExecutionException) {
+            final Throwable cause = t.getCause();
+            if (cause != null) {
+                return peel(cause, allowedType);
+            } else {
+                return new HazelcastException(t);
+            }
+        }
+
+        if (allowedType != null && allowedType.isAssignableFrom(t.getClass())) {
+            return t;
+        }
+
+        return new HazelcastException(t);
```

Comment:
However, perhaps this is not the place where to fix this: the method is already trapped into returning an unchecked exception. We'll have to make `InterruptedException` a special case earlier on and we'll probably need to separately cover the cases where an `InterruptedException` may have occurred and where not.